# Telecom360 AI

## Sprint 1 - Data Discovery

### Objective
Understand the IBM Telco Customer Churn dataset before preprocessing and model development.

### Business problem
The Telecom company wants to identify customers who are likely to churn and understand the factors contributing to churn.

### Expected Outcome
- Understand the data structure.
- Identify the target variable.
- Identify feature types.
- Detect missing values and data quality issues.

### Enginnering Notes
This notebook performs dataset discovery only.
No preprocessing.
No model training.



In [1]:
import pandas as pd
import numpy as np



In [2]:
df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.shape

(7043, 33)

### Observation

- The dataset contains **7,043 customer records** and **33 columns**.
- The dataset size is suitable for analysis using Pandas.
- No distributed processing framework is required.

In [3]:
df.columns.tolist()

['CustomerID',
 'Count',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Lat Long',
 'Latitude',
 'Longitude',
 'Gender',
 'Senior Citizen',
 'Partner',
 'Dependents',
 'Tenure Months',
 'Phone Service',
 'Multiple Lines',
 'Internet Service',
 'Online Security',
 'Online Backup',
 'Device Protection',
 'Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charges',
 'Total Charges',
 'Churn Label',
 'Churn Value',
 'Churn Score',
 'CLTV',
 'Churn Reason']

### Observation

- The dataset contains customer demographic information, service subscriptions, billing details, geographic information, CLTV, and churn-related variables.

- The dataset also includes several columns that may cause target leakage and will require further investigation.

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

In [5]:
df.head(10)

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
5,4190-MFLUW,1,United States,California,Los Angeles,90020,"34.066367, -118.309868",34.066367,-118.309868,Female,...,Month-to-month,No,Credit card (automatic),55.20,528.35,Yes,1,78,5925,Competitor offered higher download speeds
6,8779-QRDMV,1,United States,California,Los Angeles,90022,"34.02381, -118.156582",34.023810,-118.156582,Male,...,Month-to-month,Yes,Electronic check,39.65,39.65,Yes,1,100,5433,Competitor offered more data
7,1066-JKSGK,1,United States,California,Los Angeles,90024,"34.066303, -118.435479",34.066303,-118.435479,Male,...,Month-to-month,No,Mailed check,20.15,20.15,Yes,1,92,4832,Competitor made better offer
8,6467-CHFZW,1,United States,California,Los Angeles,90028,"34.099869, -118.326843",34.099869,-118.326843,Male,...,Month-to-month,Yes,Electronic check,99.35,4749.15,Yes,1,77,5789,Competitor had better devices
9,8665-UTDHZ,1,United States,California,Los Angeles,90029,"34.089953, -118.294824",34.089953,-118.294824,Male,...,Month-to-month,No,Electronic check,30.20,30.2,Yes,1,97,2915,Competitor had better devices


### Observation

- Most columns contain complete information.
- `Churn Reason` contains missing values, which appear to be business-driven.
- `Total Charges` is stored as a string and requires further investigation.

## Initial Findings

- Dataset contains 7043 customer records.
- No Sigificant missing values except Churn Reason.
- Missing values in Churn Reason appear business-driven.
- Total Charges is stored as text and requires investigation.
- Dataset includes customer location features.
- Dataset includes CLTV, enabling customer prioritization.
- Target leakage columns identified:
    - Churn Label
    - Churn Value
    - Churn Score
    - Churn Reason

# Investigating Data Quality

## Investigating the Count Column

In [6]:
df['Count'].value_counts()

Count
1    7043
Name: count, dtype: int64

### Observation

Every record contains the value **1**.

The column has no variation and therefore provides no predictive value.

It will be removed during preprocessing.


## Investigating the Total Charges Column

In [7]:

df[df['Total Charges'].str.strip()=='']



,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
2234,4472-LVYGI,1,United States,California,San Bernardino,92408,"34.084909, -117.258107",34.084909,-117.258107,Female,...,Two year,Yes,Bank transfer (automatic),52.55,,No,0,36,2578,NaN
2438,3115-CZMZD,1,United States,California,Independence,93526,"36.869584, -118.189241",36.869584,-118.189241,Male,...,Two year,No,Mailed check,20.25,,No,0,68,5504,NaN
2568,5709-LVOEQ,1,United States,California,San Mateo,94401,"37.590421, -122.306467",37.590421,-122.306467,Female,...,Two year,No,Mailed check,80.85,,No,0,45,2048,NaN
2667,4367-NUYAO,1,United States,California,Cupertino,95014,"37.306612, -122.080621",37.306612,-122.080621,Male,...,Two year,No,Mailed check,25.75,,No,0,48,4950,NaN
2856,1371-DWPAZ,1,United States,California,Redcrest,95569,"40.363446, -123.835041",40.363446,-123.835041,Female,...,Two year,No,Credit card (automatic),56.05,,No,0,30,4740,NaN
4331,7644-OMVMY,1,United States,California,Los Angeles,90029,"34.089953, -118.294824",34.089953,-118.294824,Male,...,Two year,No,Mailed check,19.85,,No,0,53,2019,NaN
4687,3213-VVOLG,1,United States,California,Sun City,92585,"33.739412, -117.173334",33.739412,-117.173334,Male,...,Two year,No,Mailed check,25.35,,No,0,49,2299,NaN
5104,2520-SGTTA,1,United States,California,Ben Lomond,95005,"37.078873, -122.090386",37.078873,-122.090386,Female,...,Two year,No,Mailed check,20.00,,No,0,27,3763,NaN
5719,2923-ARZLG,1,United States,California,La Verne,91750,"34.144703, -117.770299",34.144703,-117.770299,Male,...,One year,Yes,Mailed check,19.70,,No,0,69,4890,NaN
6772,4075-WKNIU,1,United States,California,Bell,90201,"33.970343, -118.171368",33.970343,-118.171368,Female,...,Two year,No,Mailed check,73.35,,No,0,44,2342,NaN


### Observation

A total of **11 records** contain blank values in `Total Charges`.

These blank values explain why Pandas imported the column as a string.

The issue will be handled during Sprint 3 (Data Preprocessing).

In [8]:
data_dictionary=pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str)
})
data_dictionary

,Column,Data Type
CustomerID,CustomerID,str
Count,Count,int64
Country,Country,str
State,State,str
City,City,str
Zip Code,Zip Code,int64
Lat Long,Lat Long,str
Latitude,Latitude,float64
Longitude,Longitude,float64
Gender,Gender,str


### Data Dictionary Notes

The initial data dictionary has been generated to document every feature in the dataset.

During preprocessing, each column will be classified as:

- Feature
- Target
- Identifier
- Metadata
- Target Leakage

This documentation will guide feature selection before model training.


# Sprint 1 Summary

## Completed

- Loaded the dataset successfully.
- Explored dataset dimensions.
- Reviewed all dataset columns.
- Examined data types.
- Identified missing values.
- Investigated data quality issues.
- Generated the initial data dictionary.

## Key Findings

- Dataset contains **7,043 telecom customers**.
- `CustomerID` is the unique identifier.
- `Churn Value` will be used as the target variable.
- `Count` provides no predictive value.
- `Total Charges` contains 11 blank records.
- `Churn Label`, `Churn Score`, and `Churn Reason` represent target leakage.
- `CLTV` can be used for customer prioritization.